util/ranger/new_refiner.go

```diff
@@ -243,6 +244,50 @@ func BuildRange(sc *variable.StatementContext, conds []expression.Expression, ra
 	return
 }
 
+// BuildTempRange is a method which can temporally calculate IntColumnRange, ColumnRange, IndexRange
+// with evaulating a parameterized access condition.
+// TODO: should consolidate with BuildRange
+func BuildTempRange(ctx context.Context, schema *expression.Schema, conds []expression.Expression, rangeType int, cols []*expression.Column, lengths []int) (retRanges []types.Range,
+	accessConditions, otherConditions []expression.Expression, _ error) {
+	sc := ctx.GetSessionVars().StmtCtx
+	if rangeType == IntRangeType {
+		accessConditions, otherConditions = DetachColumnConditions(conds, cols[0].ColName)
+		tmpAccessConditions := expression.EvaluateDeferredExprs(ctx, schema, accessConditions)
+		ranges, err := BuildTableRange(tmpAccessConditions, sc)
+		if err != nil {
+			return nil, nil, nil, errors.Trace(err)
+		}
+		retRanges = make([]types.Range, 0, len(ranges))
+		for _, ran := range ranges {
+			retRanges = append(retRanges, ran)
+		}
+	} else if rangeType == ColumnRangeType {
+		accessConditions, otherConditions = DetachColumnConditions(conds, cols[0].ColName)
+		tmpAccessConditions := expression.EvaluateDeferredExprs(ctx, schema, accessConditions)
+		ranges, err := buildColumnRange(tmpAccessConditions, sc, cols[0].RetType)
+		if err != nil {
+			return nil, nil, nil, errors.Trace(err)
+		}
+		retRanges = make([]types.Range, 0, len(ranges))
+		for _, ran := range ranges {
+			retRanges = append(retRanges, ran)
+		}
+	} else if rangeType == IndexRangeType {
+		var eqAndInCount int
+		accessConditions, otherConditions, _, eqAndInCount = detachIndexScanConditions(conds, cols, lengths)
+		tmpAccessConditions := expression.EvaluateDeferredExprs(ctx, schema, accessConditions)
```

Comment:
`accessConditions` only contains functions `EQ(Column, Constant)`, no need to call `expression.EvaluateDeferredExprs()` here

Reply:
No. when we use the plan cache for prepared statements, accessConditions can contain "getparam()" expressions. This is because, in `expr_to_pb.go: ExpressionsToPB()`, the deferred expressions, e.g., "getparam()", are considered as the expressions pushed, but the deferred ones are evaluated when (re)generating PB expressions.

For an example query of YCSB, "select * from usertable where ycsb_key = ?", since "ycsb_key" is the primary key of "user_table", the index-scan over "ycsb_key" is involed and accessContitions of the plan cached contains "getparam()" expressions. 

`30 func ExpressionsToPB(sc *variable.StatementContext, exprs []Expression, client kv.Client) (pbExpr *tipb.Expr, pushed []Expression, remained []Expression) {`
`31     pc := pbConverter{client: client, sc: sc}`
`32     for _, expr := range exprs {`
`33         v := pc.exprToPB(expr)`
`34         if v == nil {`
`35             if IsDeferredExpr(expr) {`
`36                 pushed = append(pushed, expr)`
`37             } else {`
`38                 remained = append(remained, expr)`
`39             }`
`40             continue`
`41         }`
`42         pushed = append(pushed, expr)`
`43         if pbExpr == nil {`
`44             pbExpr = v`
`45         } else {`
`46             // Merge multiple converted pb expression into a CNF.`
`47             pbExpr = &tipb.Expr{`
`48                 Tp:       tipb.ExprType_And,`
`49                 Children: []*tipb.Expr{pbExpr, v}}`
`50         }`
51     }
52     return
53 }

Reply:
`getEQColOffset` has been rewritten to include deferred constants.